lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -532,6 +532,37 @@ LogicalResult CircuitOp::verifyRegions() {
       mlir::SymbolTable::Visibility::Public)
     return mainModule->emitError("main module must be public");
 
+  // Check enable/disable layers arrays.
+  if (auto enabledLayers = getEnableLayersAttr())
+    for (auto enabledLayer : enabledLayers.getAsRange<SymbolRefAttr>()) {
+      auto *op = symtbl.lookupSymbolIn(*this, enabledLayer);
+      if (!op)
+        return emitOpError("unknown layer ")
+               << enabledLayer << " in enable layers list";
+      if (!isa<LayerOp>(op))
+        return emitOpError("expected layer for ")
+            .append(enabledLayer, " in enable layers list")
+            .attachNote(op->getLoc())
+            .append("not a layer");
+    }
+  if (auto disabledLayers = getDisableLayersAttr())
+    for (auto disabledLayer : disabledLayers.getAsRange<SymbolRefAttr>()) {
+      auto *op = symtbl.lookupSymbolIn(*this, disabledLayer);
+      if (!op)
+        return emitOpError("unknown layer ")
+               << disabledLayer << " in disable layers list";
+      if (!isa<LayerOp>(op))
+        return emitOpError("expected layer for ")
+            .append(disabledLayer, " in disable layers list")
+            .attachNote(op->getLoc())
+            .append("not a layer");
+      if (llvm::is_contained(cast<FModuleLike>(mainModule).getLayers(),
+                             disabledLayer))
+        return mainModule->emitOpError(
+                   "main module has circuit-disabled layer ")
+               << disabledLayer << " enabled";
+    }
```

Comment:
Two questions (I'm not sure of the answers):

1. Is there a performance issue with the use of `lookupSymbolIn` here where this would compute the symbol table multiple times?
2. Are the enable/disable layer circuit options and enable layer module level option tracked as symbol users? If they are, then would this be more suited to symbol user verifier on the layer operations?

Reply:
Thanks!

Regarding (1) -- looks like maybe yes, whoops.  Boooo (but thanks!).
Forgot that the efficient version is part of SymbolTableCollection via verifySymbolUses, I'll check if can leverage the SymbolUserOpInterface on the circuitOp itself or not (I'd guess not, but let's find out -- not w.r.t. symbols defined within itself anyway, but I'll check (for example I don't think a SymbolTable operation is itself considered a user of symbols within itself -- symbol references it has should resolve in parent SymbolTable).

If it's lazily built, maybe can be used here, since we already built the SymbolTable above so the extra steps are unlikely to be a problem re:nested.

Regarding (2), yes and we do that partially today.  I have PR in-flight for that (cc #7447), I'll poke at this a bit more.
That doesn't get us (1) however.

I'll submit that PR and update this one in a few, thanks!

Reply:
Just making sure we're on the same page, switching `symtbl.lookupSymbolIn(*this, disabledLayer)` to `symtbl.lookup(disabledLayer)` should fix 1 if the other stuff doesn't work out.

Reply:
I forgot that this is a nested symbol reference - so that doesn't actually work - woops

Reply:
Verifying enable lists on modules, including checking for enable of circuit-disabled layers here: https://github.com/llvm/circt/pull/7557 .